A CDC replication test.

kbr- opened the pull request:
This test sets up two clusters, runs a workload on one of them (the "master"), and runs an application ("replicator") which uses CDC to copy data from the master to the other cluster (the "replica") as it appears on the master.

It uses two external tools:
1. A variant of [scylla-migrate](https://github.com/scylladb/scylla-migrate/pull/15),
2. @haaawk's [CDC replicator](https://github.com/haaawk/scylla-cdc-java).

I host a version of each tool in an AWS bucket.

The PR also adds a jenkinsfile containing a Jenkins pipeline to run the test.

roydahan commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+
+class CDCReplicationTest(ClusterTester):
+    def test_replication(self):
+        print("Adding node to db_cluster... ")
+        node2 = self.db_cluster.add_nodes(count=1, enable_auto_bootstrap=True)[0]
+        print("Waiting for added node to initialize...")
+        self.db_cluster.wait_for_init(node_list=[node2])
+
+        print("Adding node to db_cluster... ")
+        node3 = self.db_cluster.add_nodes(count=1, enable_auto_bootstrap=True)[0]
+        print("Waiting for added node to initialize...")
+        self.db_cluster.wait_for_init(node_list=[node3])
```
why don't you create the db_cluster with initial size of 3?

kbr- replied:
Because SCT starts the nodes with `auto_bootstrap=off` (https://github.com/scylladb/scylla-cluster-tests/issues/2247), and CDC won't work properly if the nodes aren't bootstrapped properly.

roydahan replied:
If this really the case and we must start empty cluster with bootstrap = true, we will change that.
No need to handle it specifically in the test.

kbr- replied:
It should at least be configurable.

fruch replied:
it's should default to True, and seem like we are failing to do so.
and on initial node boot it's False, and only later we change it to True.

I thought we already cleared that in SCT code, I was wrong...

roydahan commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+        master_node = self.db_cluster.nodes[0]
+        replica_node = self.cs_db_cluster.nodes[0]
+
+        print('Creating table on master cluster.')
```
Better to use our logger.info

roydahan commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+            sess.execute(keyspace_def(1))
+            sess.execute(TABLE_DEF_BASE)
+
+        print('Waiting 40s for the latest CDC generation to start...')
```
What is the 40s timing?

kbr- replied:
When a node bootstraps, it starts a new "CDC generation", which has an associated timestamp, denoting "the moment it starts operating". A CDC generation contains a list of streams which is used when writing to CDC-enabled tables. When a write is performed to such a table, a second write is performed to its associated CDC log table; the partition key of this second write is a stream identifier taken from the CDC generation that is currently operating (based on the coordinator's local clock). More details: https://docs.scylladb.com/using-scylla/cdc/cdc-stream-generations/.

The timestamp of a new generation created by a bootstrapping node is chosen by taking the node's clock and adding 30 seconds.

I want every coordinator's clock to get past the latest generation's timestamp (the one created by `node3`), so I'm waiting 40 seconds (so I assume a clock drift of less than 10 seconds).

roydahan commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+        loader_node : BaseNode = self.loaders.nodes[0]
+
+        print('Installing tmux on loader node.')
+        res = loader_node.remoter.run(cmd='sudo yum install -y tmux')
```
We can add that to loader setup if that's really needed.

kbr- replied:
There are probably other ways to achieve the effect that I'm achieving using tmux, but I know and like tmux so I used it in this test.

So if it's not a problem, it would be nice to have tmux in the setup (it's a small utility so shouldn't be a problem).

fruch replied:
I like tmux alot too, and there shouldn't be a problem to install it. just move it to the `BaseLoaderSet.node_setup()`, I think you can add it next to `screen` install
 
but for the long run, we usually care about the output of those kind of processes, and in lots of case we take metrics out of it, and search for error on it during runtime.

kbr- replied:
> but for the long run, we usually care about the output of those kind of processes, and in lots of case we take metrics out of it, and search for error on it during runtime.

I wonder what would be the best way to get output "in real time". Spawn a thread and poll for the output?

bentsi replied:
Why didn't you use screen/nohup? 
Anyway, I think we must not install additional tools that are test specific for every run that we have, so let's keep it here.

kbr- replied:
I never used screen/nohup, I always used tmux :)

roydahan commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+        time.sleep(30)
+
+        print("Stopping nemesis before bootstrapping a new node.")
+        self.db_cluster.stop_nemesis(timeout=60)
```
if you plan to use other nemesis than RandomInterruptionNetworkMonkey, you may need to wait for the nemesis to actually finish.

kbr- replied:
I saw in the code that `stop_nemesis` joins all nemesis threads. What else do I need to do?

fruch replied:
this is o.k. usually we want more time, since some Nemesis might take longer then 60sec
I think we wait 30min.

roydahan commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+
+        with self.cql_connection_patient(node=master_node) as sess:
+            print('Fetching master table...')
+            res = sess.execute(SimpleStatement('select * from ks.tb',
```
Does it work without timing out?

kbr- replied:
It uses paging to perform the query. If I understand correctly, multiple round-trips to the cluster are performed underneath.

A bunch of local runs from my laptop didn't timeout here.

roydahan commented on test-cases/cdc/cdc-15m-replication.yaml:
```diff
@@ -0,0 +1,38 @@
+test_duration: 3600
```
Many of the params here aren't really needed as they are part of the defaults or will be overriden as part of the pipeline file.
I'll help you to clean it after you have a pipeline file.

bentsi replied:
yes need to clean the unneeded params that are part of defaults/defined in Jenkins pipeline

roydahan commented on test-cases/cdc/cdc-15m-replication.yaml:
```diff
@@ -0,0 +1,38 @@
+
+n_monitor_nodes: 0
+
+nemesis_class_name: 'RandomInterruptionNetworkMonkey'
```
May I ask why specifically this nemesis?

kbr- replied:
From my knowledge of CDC internals I believe that CDC is a feature that can potentially be very susceptible to network partitions.

fruch replied:
This nemesis won't actually run unless configured in a specific way:
```
        if not self.cluster.extra_network_interface:
            raise UnsupportedNemesis("for this nemesis to work, you need to set `extra_network_interface: True`")
```
it would configure 2 network interfaces, and Scylla should be working with one of them, and the test control with a different one. FYI it a nemesis that would work only in AWS, and we run it a bit less frequently (i.e. not daily like other Nemesis) 

`RejectInterNodeNetworkMonkey` might be a bit more useful, but we disabled it cause of "reasons"

kbr- replied:
I did enable the option:
```

# Required by the nemesis:
extra_network_interface: true 
```

What is the difference between `RandomInterruptionNetworkMonkey` and `RejectInterNodeNetworkMonkey`? Names suggest that the first is a superset of the second.

fruch replied:
the difference is details, i.e. the setup and tooling
`RejectInterNodeNetworkMonkey` is using iptables and works only on specific port and effect only communication between nodes,
while `RandomInterruptionNetworkMonkey` is using [tc](https://linux.die.net/man/8/tc) on a specific interface, hence effect also the clients communication.

kbr- commented:
Requires fixing https://github.com/scylladb/scylla-cluster-tests/issues/2295.
Also, currently needs https://github.com/scylladb/scylla-cluster-tests/pull/2296, but I guess I will eventually add monitoring nodes.

aleksbykov commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+        f.flush()
+        os.fsync(f.fileno())
+
+SCYLLA_MIGRATE_URL = "https://kbr-scylla.s3-eu-west-1.amazonaws.com/scylla-migrate"
+REPLICATOR_URL = "https://kbr-scylla.s3-eu-west-1.amazonaws.com/scylla-cdc-replicator-0.0.1-SNAPSHOT-jar-with-dependencies.jar"
+
```
can this utility be moved to docker?

kbr- replied:
Eventually, but not right now: at least the replicator is in continuous developement (we're fixing bugs and adding handling of new types).

Why would we want to put it in docker?

aleksbykov replied:
it wil be easy to use them in common infrastructure. Allow to assign LogWatchers for on runtime monitoring and collect appropriate logs for future analyzes

kbr- replied:
Ok. We can put them into dockers when they mature a bit. At this time I would say it doesn't make much sense.

aleksbykov commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+        replica_node = self.cs_db_cluster.nodes[0]
+
+        print('Creating table on master cluster.')
+        with self.cql_connection_patient(node=master_node) as sess:
+            sess.execute(keyspace_def(3))
+            sess.execute(TABLE_DEF_MASTER)
```
i think it is not need on master cluster. Because cassandra will create it automatically from profile.

kbr- replied:
I actually removed the keyspace definition from the profile to reduce duplication. I wanted to have the KS definition in a single place. Since I need to have it in the code anyway (to create the KS on the second cluster), I decided to put it here.

bentsi commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,193 @@
+            print('Consistency check successful.')
+            return
+
+        # Collect more data for analysis
```
better to put in a separate method like `collect_data_for_analysis`

kbr- commented:
v2:
- add a jenkinsfile
- remove unneeded parameters from yaml
- use `self.log` instead of `print`
- other defaults (e.g. qa in email recipients)

kbr- commented:
v3:
- don't bootstrap the 2nd and 3rd node from the test code, use the new bootstrapping method implemented by Aleks
- separate collection of data for analysis into separate method
- move schema definition into cassandra-stress profile; in the code, query the schema from the master cluster to create it on the replica cluster

kbr- commented:
@roydahan @fruch @bentsi @aleksbykov please re-review

aleksbykov commented on data_dir/cdc_replication_profile.yaml:
```diff
@@ -0,0 +1,29 @@
+keyspace: ks1
+table: table1
```
Move it before table definition. otherwise c-s stress failed to create table

kbr- replied:
Actually it does create the table successfully (tested). The order of config options should not matter, if C-S starts by loading everything into memory first (I guess it does).

I can reorder it anyway to be consistent with other profiles.

aleksbykov commented on jenkins-pipelines/cdcReplicationPipeline.jenkinsfile:
```diff
@@ -0,0 +1,180 @@
+#!groovy
+
+// Import shared libraries
```
I think, better to use prepared the pipeline longevity and define jenkins file as for other jobs
And define the type and version of oracle in config yaml

kbr- replied:
Unfortunately, the existing pipelines don't suit my needs :(
- I want the test to keep the cluster if any failure happens so that I can ssh into it and investigate. The longevity pipeline only keeps the cluster on internal framework failures (that are reported "critical") AFAIK.
- I want to be able to specify the oracle version from Jenkins. I'd like to have a bunch of jobs running regularly for different Scylla versions.
- I also have a patch prepared that extends this pipeline by adding an option to use gemini instead of cassandra-stress. I decided to push it as a separate PR if that one is merged.

bentsi replied:
Longevity pipeline supports keeping the instances, you can set the `post_behavior_db_nodes` parameter:
![image](https://user-images.githubusercontent.com/5269241/87331379-8817dc80-c542-11ea-9334-bf3bafaa5b8e.png)
Regarding other things, we indeed not support since we don't have generalized parameters function

kbr- replied:
AFAIK `keep` always keeps the instances. I want to keep them only if the test fails.
So I would like to use `keep-on-failure`, however it doesn't do exactly what I want (https://github.com/scylladb/scylla-cluster-tests/issues/1825).

aleksbykov commented on cdc_replication_test.py:
```diff
@@ -0,0 +1,203 @@
+
+        self.log.info('Fetching schema definitions from master cluster.')
+        master_node = self.db_cluster.nodes[0]
+        master_cluster = Cluster([master_node.external_address])
+        master_cluster.control_connection_timeout = 20
+        # need to connect so the driver can fetch schema metadata,
+        # even though we don't need the session itself:
+        with master_cluster.connect() as _:
+            master_cluster.refresh_schema_metadata()
+            ks = master_cluster.metadata.keyspaces[self.KS_NAME]
```
I think better use standard sct methods for connection, this allow fix any connection issue without patching the test.
These will make sense, when cassandra driver will be updated and will not allow to use list of nodes in constructor for Cluster class.
for example next could be use:
```
        with self.cql_connection_patient_exclusive(node=master_node) as scylla_session:
            scylla_session.cluster.refresh_schema_metadata()
            ks = scylla_session.cluster.metadata.keyspaces[self.KS_NAME]
```

kbr- replied:
Thanks, I was wondering how to get the `Cluster` object.

kbr- commented:
v4:
- rebased
- fixed order of definitions in c-s profile
- use `self.cql_connection_patient` and access `Cluster` through the session instead of creating my own `Cluster`

kbr- commented:
@fruch @bentsi @roydahan it looks like all comments are addressed, could we merge this?

roydahan commented:
@kbr- / @aleksbykov please make sure we have a Jenkins job to trigger it in the correct place (I think should be under features).
